### ISSUE TITLE
Reimplement numpy functions as pure python

### ISSUE THREAD

**Issue by guilhermeleobas** [xnd-project/rbc#31]
## Why this is needed

Currently, there is no way to use a NumPy function with an omnisci array. Reimplement those functions would allow one to just call `np.sum(an_omnisci_array)` for instance. Furthermore, some functions/methods might also require the usage of [Numba Runtime (NRT)](https://numba.pydata.org/numba-doc/dev/developer/numba-runtime.html) extension which is probably not available in the server-side.

## An example

```python
def foo(omnisci_array):
    return np.sum(omnisci_arrray)
```

## List of functions that we must implement

This list might grow in the future.

- [ ] `np.prod(t) # t is an omnisci array.`
- [ ] `np.sum(t)`
- [ ] `np.nanprod(t)`
- [ ] `np.nansum(t)`
- [ ] `np.cumprod(t)`
- [ ] `np.cumsum(t)`
- [ ] `np.nancumprod(t)`
- [ ] `np.nancumsum(t)`
- [ ] `np.diff(t)`
- [ ] `np.ediff1d(t)`
- [ ] `np.gradient(t)`
- [ ] `np.cross(t)`
- [ ] `np.trapz(t)`

**Comment by rgommers** [comment]
I'm probably missing some context here, so not sure if this makes sense, but: have you considered adding `__array_function__` and `__array_ufunc__` to the omnisci array object? Or is that what you mean to do here, so `np.sum` dispatches to your new `sum` function that does support omnisci arrays?

**Comment by guilhermeleobas** [comment]
>  Or is that what you mean to do here, so np.sum dispatches to your new sum function that does support omnisci arrays?

That's what I mean. Provide an implementation for `sum/np.sum` for omnisci array. For instance, using the `sum` example:

```python
@numba.extending.overload(np.sum)
@numba.extending.overload(sum)
def omnisci_sum(a):
    def impl(a):
        s = 0 
        n = len(a)
        for i in range(n):
            s += a[i]
        return s
    return impl

@omni('i32(i32[])')
def bar(t):
  return np.sum(t)

omni.register()

_, result = omni.sql_execute('select t, bar(t) from math') # works
```

**Comment by pearu** [comment]
@rgommers , yes, the context here is a bit different from the context where `__array_function__` could be used.

All functions that rbc decorates, are never executed. Instead, the rbc decorated functions are used as templates for generating LLVM IR module strings that are sent to the omniscidb server. Omniscidb server does not know anything about Python nor numpy arrays. This means that the LLVM IR module must be self-contained in the sense that the functions in the module can use only LLVM intrinsic functions or functions and data structures that the omniscidb server defines. One of the omniscidb data structures is `Array` that represents a 1-d array that basically holds only the data pointer value and information about the size and type of the array.

When omniscidb `Array` is used inside the rbc decorated function as `x`, currently, only the following operations are implemented:
```
len(x)        # returns the size of Array as integer
v = x[i]      # set `v` to the value of the `i`-th Array item
x[i] = v      # set the value of the `i`-th Array item to `v`
```
Using these basic operations, one can easily implement all other operators such as `sum`, etc.

Now, when using `np.sum` inside a rbc decorated functions, it expects numba view of the numpy array as the argument to `np.sum`. To use omniscidb Array as the argument to `np.sum`, there seems to be two options:
1. Create a temporary numpy array object, copy the values of Array to the numpy object, and use the temporary array object as argument to `np.sum`:
```
@omnisci(..)
def foo(x):
    # x is omnisci Array
    # compute np.sum(x) as follows
    y = np.empty(len(x), dtype=...)
    for i in range(len(x)):
        y[i] = x[i]
    s = np.sum(y)
    ...
```
2. Provide Python implementations of numpy methods:
```
@omnisci(..)
def foo(x):
    # compute np.sum(x) as follows
    s = 0
    for i in range(len(x)):
        s += x[i]
```

My cut feeling is that option 1. is suboptimal because of it requires an unnecessary copy of the data and one needs to implement a new data type that is a omniscidb view of the numpy array.
Option 2. would involve replacing the real `numpy` namespace with an rbc namespace that implements numpy API functions as pure python functions that can be rbc decorated and therefore used from UDF definitions.

**Comment by rgommers** [comment]
Thanks both, I get the picture now. Interesting stuff!

**Comment by guilhermeleobas** [comment]
@pearu and @rgommers, with respect to the first option. Wouldn't it be possible to convert the inner structure of an OmniSci array to a NumPy array? I do not how a NumPy array is implemented under the hood but I think we can achieve something similar to what PyTorch does with [torch.from_numpy()](https://pytorch.org/docs/stable/torch.html#torch.from_numpy) and [torch.tensor.numpy()](https://pytorch.org/docs/stable/tensors.html#torch.Tensor.numpy).
